### DISCUSSION TITLE
Injecting mock data into my test function

### DISCUSSION THREAD

**Discussion by ViniciusAndradex** [asomers/mockall#576] [Questions]
Hi, I'm trying to create unit tests and I'm using mockall to create the mocks. However, I'm having trouble making my method use the mocked data instead of my database, could anyone help me? I will pass the code.

Repository
```
#[cfg_attr(test, automock)]
impl GenericRepositoryGet<Mediunship> for TestRepository {
    async fn get_all(&self) -> Result<Vec<Test>, DatabaseErrors> {
        let mut conn = self.new_connection().await?;
        let vec_itens = tests.load::<Mediunship>(&mut conn);

        match vec_itens {
            Ok(values) => Ok(values),
            Err(err) => {
                log::error!("Error: {}", err);
                Err(DatabaseErrors::FetchingDataError(err))
            }
        }
    }
```

Service
```
#[cfg_attr(test, automock)]
impl TraitTestService for TestService {
    async fn init(&self) -> TestService {
        TestService
    }

    async fn new_repository(&self) -> TestRepository {
        TestRepository.init().await
    }

    async fn get_all(&self) -> Result<Vec<ResponseTestDto>, Error> {
        let repository = self.new_repository().await;

        let test_list = repository.get_all().await?;
        let list = test_list
            .iter()
            .map(|obj| ResponseTestDto {
                id: obj.id.unwrap(),
                test: obj.test.clone(),
            })
            .collect();
        Ok(list)
    }
```

Test

    #[actix_rt::test]
    async fn test_get_all() {

        let mut mock_repository = MockTestRepository::new();
        let mut mock_service = MockTestService::new();

        mock_repository
        .expect_get_all()
        .times(1)
        .returning(|| {
            Ok(vec![create_test(
                Some(1),
                "Doutrinador".to_string(),
            )])
        });

        mock_service
            .expect_get_all()
            .times(1)
            .returning(|| Ok(vec![create_reponse_dto(1, "Doutrinador".to_string())]));

        let expect_data = mock_service.get_all().await.unwrap();
        let result = MediunshipService.get_all().await.unwrap(); // In this line I am making the instance of my service, but I don't know how to make my service receive data from my mock, how could I do this? 

        assert_eq!(
            expect_data,
            result
        );
    }
```

**Comment by asomers** [comment]
There's a lot of information missing from your post.
* What is `MediunshipService`?  It isn't defined anywhere.
* Is the test compiled with or without `#[cfg(test)]`?  That is, is it a unit test or a functional test?
* Do you want `MockTraitTestService` to be compiled with real `TestRepository` or `MockTestRepository`?
* What files or modules do all these snippets come from?  Crucially, if they're in different files, then how does the file with `TraitTestService` get `TestRepository` with `use` ?

Also, pro tip: your posts will be more readable if you use follow the suggestions at https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting to use syntax highlighting.
